File: 3296-Minimum-Number-of-Seconds-to-Make-Mountain-Height-Zero/Solution.py
# Approach 1 — Binary Search + Simulation (Python)

### Time Complexity
O((n + H) log H)

### Space Complexity
O(1)

### Code

```python
class Solution:

    def isPossible(self, time, height, worker):

        for i in range(len(worker)):

            j = 1
            curr = 0

            while height > 0 and curr + j * worker[i] <= time:
                curr += j * worker[i]
                j += 1
                height -= 1

            if height == 0:
                return True

        return height == 0


    def minNumberOfSeconds(self, mountainHeight, workerTimes):

        left = 1
        right = 0
        slowest = 1

        for t in workerTimes:
            slowest = max(slowest, t)

        for i in range(1, mountainHeight + 1):
            right += slowest * i

        while left < right:

            mid = (left + right) // 2

            if self.isPossible(mid, mountainHeight, workerTimes):
                right = mid
            else:
                left = mid + 1

        return left
```

---

# Approach 2 — Binary Search + Mathematical Trick (Python)

### Time Complexity
O(n log H)

### Space Complexity
O(1)

### Code

```python
import math

class Solution:

    def isPossible(self, time, height, worker):

        totalHeight = 0

        for w in worker:

            k = int((math.sqrt(1 + (8 * time) / w) - 1) / 2)

            totalHeight += k

            if totalHeight >= height:
                return True

        return False


    def minNumberOfSeconds(self, mountainHeight, workerTimes):

        left = 1
        right = 0
        slowest = 1

        for w in workerTimes:
            slowest = max(slowest, w)

        for i in range(1, mountainHeight + 1):
            right += slowest * i

        while left < right:

            mid = (left + right) // 2

            if self.isPossible(mid, mountainHeight, workerTimes):
                right = mid
            else:
                left = mid + 1

        return left
```

File: 3296-Minimum-Number-of-Seconds-to-Make-Mountain-Height-Zero/Solution.cpp
# Approach 1 — Binary Search + Simulation (C++)

### Time Complexity
O((n + H) log H)

### Space Complexity
O(1)

### Code

```cpp
class Solution {
public:

    bool isPossible(long long time, int height, vector<int>& worker) {

        for (int i = 0; i < worker.size(); i++) {

            long long j = 1;
            long long curr = 0;

            while (height > 0 && curr + j * worker[i] <= time) {

                curr += j * worker[i];
                j++;
                height--;
            }

            if (height == 0) return true;
        }

        return height == 0;
    }

    long long minNumberOfSeconds(int mountainHeight, vector<int>& workerTimes) {

        long long left = 1;
        long long right = 0;
        long long slowest = 1;

        for (int t : workerTimes) {
            slowest = max(slowest, (long long)t);
        }

        for (int i = 1; i <= mountainHeight; i++) {
            right += slowest * i;
        }

        while (left < right) {

            long long mid = (left + right) / 2;

            if (isPossible(mid, mountainHeight, workerTimes))
                right = mid;
            else
                left = mid + 1;
        }

        return left;
    }
};
```

---

# Approach 2 — Binary Search + Mathematical Trick (C++)

### Time Complexity
O(n log H)

### Space Complexity
O(1)

### Code

```cpp
class Solution {
public:

    bool isPossible(long long time, int height, vector<int>& worker) {

        long long totalHeight = 0;

        for (int w : worker) {

            long long k = (long long)((sqrt(1 + (8.0 * time) / w) - 1) / 2);

            totalHeight += k;

            if (totalHeight >= height)
                return true;
        }

        return false;
    }

    long long minNumberOfSeconds(int mountainHeight, vector<int>& workerTimes) {

        long long left = 1;
        long long right = 0;
        long long slowest = 1;

        for (int w : workerTimes) {
            slowest = max(slowest, (long long)w);
        }

        for (int i = 1; i <= mountainHeight; i++) {
            right += slowest * i;
        }

        while (left < right) {

            long long mid = (left + right) / 2;

            if (isPossible(mid, mountainHeight, workerTimes))
                right = mid;
            else
                left = mid + 1;
        }

        return left;
    }
};
```
